### ISSUE TITLE
Explicit calls to `std::swap` are made on user provided template types. This does not work for `C++20`.

### ISSUE THREAD

**Issue by melton1968** [boostorg/sort#72]
Starting with `C++20`, [fully specializing a function template](https://en.cppreference.com/w/cpp/language/extending_std) in the `std` namespace is undefined behavior. This precludes implementing `swap` for user-defined types in the standard namespace. You can find the reasoning in this [paper](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0551r0.pdf).

I am running into this issue while using `boost::sort` in a project using `C++20`. I have implemented a free function `swap` for my type in the type's namespace, but since the library code is explicitly calling `std::swap`, it cannot find it.  If I implement the free function in `std`, the compiler chooses to ignore it, which is allowed under the standard.

I changed the sort code to use the "`use std::swap` and then call `swap` unqualified" idiom, and everything worked as expected. For example:
```c++
std::swap(*iter1, *iter2);
```
becomes,
```c++
using std::swap;
swap(*iter1, *ite2);
```

I can submit a PR for this change if that is appropriate.

**Comment by spreadsort** [comment]
On Tue, Jun 6, 2023, 5:04 PM Mark Melton ***@***.***> wrote:

> Starting with C++20, fully specializing a function template
> <https://en.cppreference.com/w/cpp/language/extending_std> in the std
> namespace is undefined behavior. This precludes implementing swap for
> user-defined types in the standard namespace. You can find the reasoning in
> this paper
> <https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0551r0.pdf>.
>
> I am running into this issue while using boost::sort in a project using
> C++20. I have implemented a free function swap for my type in the type's
> namespace, but since the library code is explicitly calling std::swap, it
> cannot find it. If I implement the free function in std, the compiler
> chooses to ignore it, which is allowed under the standard.
>
> I changed the sort code to use the "use std::swap and then call swap
> unqualified" idiom, and everything worked as expected. For example:
>
> std::swap(*iter1, *iter2);
>
> becomes,
>
> using std::swap;swap(*iter1, *ite2);
>
> I can submit a PR for this change if that is appropriate.
>
Yes, please.

**Comment by nigels-com** [comment]
Resolved in PR #73, by the look of it.

**Comment by spreadsort** [comment]
Thanks Nigel!
